Projects/Server/Json/Converters/FlagsConverter.cs

```diff
@@ -0,0 +1,71 @@
+/*************************************************************************
+ * ModernUO                                                              *
+ * Copyright 2019-2020 - ModernUO Development Team                       *
+ * Email: hi@modernuo.com                                                *
+ * File: TimeSpanConverter.cs                                            *
+ *                                                                       *
+ * This program is free software: you can redistribute it and/or modify  *
+ * it under the terms of the GNU General Public License as published by  *
+ * the Free Software Foundation, either version 3 of the License, or     *
+ * (at your option) any later version.                                   *
+ *                                                                       *
+ * You should have received a copy of the GNU General Public License     *
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>. *
+ *************************************************************************/
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Server.Json
+{
+    public class FlagsConverter<T> : JsonConverter<T> where T : struct
+    {
+        public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            var flags = 0x0;
+
+            while (true)
+            {
+                reader.Read();
+                if (reader.TokenType == JsonTokenType.EndObject)
+                {
+                    break;
+                }
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    throw new JsonException("Invalid Json structure for Flag object");
+                }
+
+                var key = reader.GetString();
+
+                reader.Read();
+
+                var val = reader.GetBoolean();
+
+                if (val)
+                {
+                    Enum.TryParse<T>(key, out var flag);
+                    flags |= (int) (object) flag;
+                }
+            }
+
+            return (T) (object) flags;
+        }
+
+        public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
+        {
+            writer.WriteStartObject();
+            foreach (var flagName in Enum.GetNames(typeof(T)))
+            {
+                if (!flagName.StartsWith("Expansion"))
```

Comment:
What is special about `Expansion` that we have to skip it?

Reply:
The expansion flags are just aggregates of the individual flags so I don't want to write those flags out to booleans

Reply:
So let's do `Enum.GetValues(typeof(T))` to get the values in an array.
Then let's check it generically:
```cs
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int GetUnderlyingTypeLength(TypeCode typeCode) =>
            typeCode switch
            {
                TypeCode.Byte   => 1,
                TypeCode.SByte  => 1,
                TypeCode.Int16  => 2,
                TypeCode.UInt16 => 2,
                TypeCode.Char   => 2,
                TypeCode.Int32  => 4,
                TypeCode.UInt32 => 4,
                TypeCode.Int64  => 8,
                TypeCode.UInt64 => 8,
                _               => 8
            };
```

```cs
var size = GetUnderlyingTypeLength(underlyingType) - 1;
...
var flag = ConvertToUInt64(underlyingType, flagValues[i]);
if (flag == 0 || (flag & size) == 0)
{
  // write value
}
```